File: 207.Course Schedule/Solution.py
import collections


class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        if not prerequisites:
            return True
        Adj = {}
        for i in range(numCourses):
            Adj[i] = []
        for edge in prerequisites:
            Adj[edge[1]].append(edge[0])
        V = numCourses*[0]
        for i in range(numCourses):
            if V[i] == 0:
                if not self.DFSvisit(i, Adj, V):
                    return False
        return True

    def DFSvisit(self, u, Adj, V):
        V[u] = 1
        for v in Adj[u]:
            if V[v] == 1:
                return False
            if V[v] == 0:
                if not self.DFSvisit(v, Adj, V):
                    return False
        V[u] = 2
        return True

File: 207.Course Schedule/src/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> courses = new ArrayList<List<Integer>>();
        for (int i = 0; i < numCourses; i++) {
        	courses.add(new ArrayList<Integer>());
        }
        for (int i = 0; i < prerequisites.length; i++) {
        	courses.get(prerequisites[i][0]).add(prerequisites[i][1]);
        }
        int[] track = new int[numCourses];
        for (int i = 0; i < numCourses; i++) {
        	if (dfs(courses, i, track) == false)
        		return false;
        }
        return true
    }
    
    public boolean dfs(List<List<Integer>> courses, int start, int[] track)
    {
    	if (track[start] == 2)
    		return true;
    	if (track[start] == 1)
    		return false;
    	track[start] = 1;
    	for (int j = 0; j < courses.get(start).size(); j++) {
    		if (dfs(courses, courses.get(start).get(j), track) == false)
    			return false;
    	}
    	track[start] = 2;
    	return true;
    }
}